core/shared/src/main/scala/fs2/async/mutable/Queue.scala

```diff
@@ -160,20 +181,50 @@ object Queue {
         def timedEnqueue1(a: A, timeout: FiniteDuration, scheduler: Scheduler): F[Boolean] = offer1(a)
         def offer1(a: A): F[Boolean] =
           qref.modify { s =>
-            if (s.deq.isEmpty) s.copy(queue = s.queue :+ a)
-            else s.copy(deq = s.deq.tail)
+            if (s.deq.isEmpty) s.copy(queue = s.queue :+ a, peek = None)
+            else s.copy(deq = s.deq.tail, peek = None)
           }.flatMap { c =>
-            if (c.previous.deq.isEmpty) // we enqueued a value to the queue
-              signalSize(c.previous, c.now).as(true)
-            else // queue was empty, we had waiting dequeuers
-              c.previous.deq.head.setAsyncPure(Chunk.singleton(a)).as(true)
+            val dq = if (c.previous.deq.isEmpty) {
+              // we enqueued a value to the queue
+              signalSize(c.previous, c.now)
+            } else {
+              // queue was empty, we had waiting dequeuers
+              c.previous.deq.head.setAsyncPure(Chunk.singleton(a))
+            }
+            val pk = if (c.previous.peek.isEmpty) {
+              // no peeker to notify
+              F.unit
+            } else {
+              // notify peekers
+              c.previous.peek.get.setAsyncPure(a)
+            }
+            (dq *> pk).as(true)
           }
 
         def dequeue1: F[A] = cancellableDequeue1.flatMap { _._1 }
 
+        def peek1: F[A] = peek1Impl.flatMap(_.fold(_.get, F.pure))
+
         def timedDequeue1(timeout: FiniteDuration, scheduler: Scheduler): F[Option[A]] =
           timedDequeueBatch1(1, timeout, scheduler).map(_.map(_.strict.head))
 
+        def timedPeek1(timeout: FiniteDuration, scheduler: Scheduler): F[Option[A]] = {
+          peek1Impl.flatMap {
+            case Left(ref) => ref.timedGet(timeout, scheduler)
+            case Right(a) => F.pure(Some(a))
+          }
+        }
+
+        private def peek1Impl: F[Either[Ref[F, A], A]] = ref[F, A].flatMap { ref =>
+          qref.modify { state =>
+            if (state.queue.isEmpty && state.peek.isEmpty) state.copy(peek = Some(ref))
+            else state
+          }.map { change =>
+            if (change.previous.queue.isEmpty) Left(change.now.peek.get)
+            else Right(change.now.queue.head)
```

Comment:
One super minor nitpick (sorry). `change.previous.queue.head` for symmetry with the `isEmpty` (since the queue hasn't changed between previous and now)

Reply:
Fair enough.